zenoh/src/api/bytes.rs

```diff
@@ -2094,12 +2222,464 @@ where
+
+    fn deserialize(self, bytes: Self::Input<'_>) -> Result<(A, B, C, D), Self::Error> {
+        let codec = Zenoh080::new();
+        let mut reader = bytes.0.reader();
+
+        let abuf: ZBuf = codec
+            .read(&mut reader)
+            .map_err(|e| ZBytesReadErrorTuple4::One(ZBytesReadError::Read(e)))?;
+        let apld = ZBytes::new(abuf);
+        let a = A::try_from(apld)
+            .map_err(|e| ZBytesReadErrorTuple4::One(ZBytesReadError::Deserialize(e)))?;
+
+        let bbuf: ZBuf = codec
+            .read(&mut reader)
+            .map_err(|e| ZBytesReadErrorTuple4::Two(ZBytesReadError::Read(e)))?;
+        let bpld = ZBytes::new(bbuf);
+        let b = B::try_from(bpld)
+            .map_err(|e| ZBytesReadErrorTuple4::Two(ZBytesReadError::Deserialize(e)))?;
+
+        let cbuf: ZBuf = codec
+            .read(&mut reader)
+            .map_err(|e| ZBytesReadErrorTuple4::Three(ZBytesReadError::Read(e)))?;
+        let cpld = ZBytes::new(cbuf);
+        let c = C::try_from(cpld)
+            .map_err(|e| ZBytesReadErrorTuple4::Three(ZBytesReadError::Deserialize(e)))?;
+
+        let dbuf: ZBuf = codec
+            .read(&mut reader)
+            .map_err(|e| ZBytesReadErrorTuple4::Four(ZBytesReadError::Read(e)))?;
+        let dpld = ZBytes::new(dbuf);
+        let d = D::try_from(dpld)
+            .map_err(|e| ZBytesReadErrorTuple4::Four(ZBytesReadError::Deserialize(e)))?;
+
+        Ok((a, b, c, d))
+    }
+}
+
+impl<A, B, C, D> TryFrom<ZBytes> for (A, B, C, D)
+where
+    A: TryFrom<ZBytes>,
+    <A as TryFrom<ZBytes>>::Error: Debug,
+    B: TryFrom<ZBytes>,
+    <B as TryFrom<ZBytes>>::Error: Debug,
+    C: TryFrom<ZBytes>,
+    <C as TryFrom<ZBytes>>::Error: Debug,
+    D: TryFrom<ZBytes>,
+    <D as TryFrom<ZBytes>>::Error: Debug,
+{
+    type Error = ZBytesReadErrorTuple4<A, B, C, D>;
+
+    fn try_from(value: ZBytes) -> Result<Self, Self::Error> {
+        ZSerde.deserialize(&value)
+    }
+}
+
+impl<A, B, C, D> TryFrom<&ZBytes> for (A, B, C, D)
+where
+    A: TryFrom<ZBytes>,
+    <A as TryFrom<ZBytes>>::Error: Debug,
+    B: TryFrom<ZBytes>,
+    <B as TryFrom<ZBytes>>::Error: Debug,
+    C: TryFrom<ZBytes>,
+    <C as TryFrom<ZBytes>>::Error: Debug,
+    D: TryFrom<ZBytes>,
+    <D as TryFrom<ZBytes>>::Error: Debug,
+{
+    type Error = ZBytesReadErrorTuple4<A, B, C, D>;
+
+    fn try_from(value: &ZBytes) -> Result<Self, Self::Error> {
+        ZSerde.deserialize(value)
+    }
+}
+
+impl<A, B, C, D> TryFrom<&mut ZBytes> for (A, B, C, D)
+where
+    A: TryFrom<ZBytes>,
+    <A as TryFrom<ZBytes>>::Error: Debug,
+    B: TryFrom<ZBytes>,
+    <B as TryFrom<ZBytes>>::Error: Debug,
+    C: TryFrom<ZBytes>,
+    <C as TryFrom<ZBytes>>::Error: Debug,
+    D: TryFrom<ZBytes>,
+    <D as TryFrom<ZBytes>>::Error: Debug,
+{
+    type Error = ZBytesReadErrorTuple4<A, B, C, D>;
+
+    fn try_from(value: &mut ZBytes) -> Result<Self, Self::Error> {
+        ZSerde.deserialize(&*value)
+    }
+}
+
+// HashMap
```

Comment:
Do you think we can provide something similar for Vec? Now we only support `Vec<u8>`

Reply:
Unfortunately not. Rust does not support serialization and `Vec<u8>` should have a special treatment compared to `Vec<T> where ZSerde: Serialize<T>`.
An iterator of `T` is serialized in this way:
- Serialize `T` as standalone
- Write length of serialized `T` into final buffer
- Append serialized `T`

This allows to iterate over the elements `T` because we know where the serialized `T` starts and end.
`u8` is a special case of `T` where we do not want to prepend the serialized length but just copy everything.
But in order to treat it differently, we need specialization in Rust... which is not supported :(